verifier.py

```diff
@@ -1,31 +1,46 @@
-import json, codecs
+import json
+import codecs
 from pages import ProjectDashboardPage, ProjectFilesPage, ProjectAnalyticsPage, \
     ProjectForksPage, ProjectRegistrationsPage, ProjectWikiPage, RegistrationDashboardPage, RegistrationFilesPage, \
     RegistrationAnalyticsPage, RegistrationForksPage, RegistrationWikiPage, UserProfilePage, InstitutionDashboardPage
 from crawler import Crawler
+import bs4
 
 
 # Verifier superclass
 class Verifier:
     def __init__(self, min_size, pg_type, end):
+        """
+        :param min_size: File size minimum for a page. Anything below this couldn't possibly be a complete file.
+        :param pg_type: The class to instantiate page objects with.
+        :param end: Indentifier in the URL, e.g. 'files/', 'end' is a misnomer ('wiki/' in the middle of a URL)
+        """
         self.minimum_size = min_size
         self.page_type = pg_type
         self.url_end = end
-        self.page_elements = {}
-        self.pages = []
+
+        # Certain elements will be absent if there's no content for them to display, so we check if there is a loading
+        # bar in its place. This means the element should exist, but it doesn't.
+        self.loading_elements = {}
+
+        # Other elements will be replaced by a message if there's no content for them (e.g. "This user has no projects")
+        # We check for the elements and their alternates if the original isn't found.
+        self.alternate_elements = {}
+
+        self.pages = []  # All the page objects
         self.failed_pages = []
 
     # Populate self.pages with the relevant files
-    def harvest_pages(self, json_filename, json_list):
+    def harvest_pages(self, json_dictionary, json_list):
         """
-        :param json_filename: The json file
+        :param json_dictionary: The dictionary created from the json file
         :param json_list: The list in the json file of found URLs
         :return: Null, but self.pages is populated.
         """
-        for url in json_list:
+        for url in json_list[:]:
```

Comment:
This is the really important addition for some reason, which allows iteration to not skip when removing.